File: CCcodes/driverFULL_monthly.m
clear
format long

site='HCKM';

fileMES1='01-Nov-2009_31-Jul-2023_HCKM_daily.mat';

%load Mesonet, Climate Data and Thresholds
dircli1='/Users/erappin/Documents/Mesonet/ClimateIndices/cliSITES/';
dircli=strcat(strcat(dircli1,site,'/'));
filethresh=strcat(dircli,strcat(site,"_CLIthresh_daily"));
load(filethresh);
filemes=strcat(dircli1,fileMES1);
load(filemes);

%TT=TT_dailyMES;
TIME=TT_dailyMES.TimestampCollected;
sTIME=TIME(1);
eTIME=TIME(end);
[sY,~,~]=ymd(sTIME);
[eY,~,~]=ymd(eTIME);
Yvalues = (datetime(sY,01,01):days(1):datetime(eY,12,31))';
TT_dailyMES = retime(TT_dailyMES,Yvalues,'fillwithconstant','Constant',0);
%isD=find(all(ismember(TT_dailyMES.TimestampCollected,sTIME),2));
%ieD=find(all(ismember(TT_dailyMES.TimestampCollected,eTIME),2));

TIME_full=TT_dailyMES.TimestampCollected;
isD=find(all(ismember(TIME_full,sTIME),2));
ieD=find(all(ismember(TIME_full,eTIME),2));


YEAR=unique(TIME_full.Year);
myStructM.year=YEAR;
MNTH=["Jan" "Feb" "Mar" "Apr" "May" "Jun" "Jul" "Aug" "Sep" "Oct" "Nov" "Dec"]';
myStructM.month=MNTH;
VAR=["TAIR_month" "DWPT_month" "TAIRx_month" "TAIRn_month" "PRCP_month" "RELH_month" ...
     "PRES_month" "SM02_month" "WDIR_month" "WSPD_month" "WSMX_month" "SRAD_month"]';
myStructM.var=VAR;

ny=numel(YEAR);
nm=numel(MNTH);
nv=numel(VAR);

Md=cell(nm,nv);
My=cell(nm,nv);

yoD = year(TIME_full);         
moD = month(TIME_full);        

for i=1:nm
    im = ismember(moD,i);
    TIME_month=TIME_full(im,:);
    for j=1:nv
        TT=getData_month(TT_dailyMES,im,TIME_full);
        TTcal=TT.TIME_month;
        TTyears = year(TTcal);
        [~,TTindYears] = findgroups(TTyears);
        Vartt=TT{:,j};
        TTbyYear = arrayfun(@(y){Vartt(TTyears==y,:)},TTindYears);
        for k = 1:numel(TTbyYear)
            TTy{:,k}=TTbyYear{k};
        end
        Md{i,j}=TTy;
        My{i,j}=TTindYears;
    end
end

myStructM.year=My;
myStructM.data=Md;

DATA=myStructM.data;
MNTH=myStructM.month;
YEAR=string(myStructM.year{1,1});

IndicesY=ones(ny,1);
for i=1:ny
    yearT(i)=datetime(double(YEAR(i)),01,01);
    IndicesY(i)=datefind(yearT(i),TIME_full);
end

return

%%%%%%%%%%%%%%%%%%function for monthly data%%%%%%%%%%%%%%%%%%%%%
function [TT_month]=getData_month(TT,mo,TIME)

TIME_month=TIME(mo,:);

TAIR_M=TT.TAIR;
TAIR_month=TAIR_M(mo,:);
TT_TAIR_mo=timetable(TIME_month,TAIR_month);

DWPT_M=TT.DWPT;
DWPT_month=DWPT_M(mo,:);
TT_DWPT_mo=timetable(TIME_month,DWPT_month);

TAIRx_M=TT.TAIRx;
TAIRx_month=TAIRx_M(mo,:);
TT_TAIRx_mo=timetable(TIME_month,TAIRx_month);

TAIRn_M=TT.TAIRn;
TAIRn_month=TAIR_M(mo,:);
TT_TAIRn_mo=timetable(TIME_month,TAIRn_month);

PRCP_M=TT.PRCP;
PRCP_month=PRCP_M(mo,:);
TT_PRCP_mo=timetable(TIME_month,PRCP_month);

RELH_M=TT.RELH;
RELH_month=RELH_M(mo,:);
TT_RELH_mo=timetable(TIME_month,RELH_month);

PRES_M=TT.PRES;
PRES_month=PRES_M(mo,:);
TT_PRES_mo=timetable(TIME_month,PRES_month);

SM02_M=TT.SM02;
SM02_month=SM02_M(mo,:);
TT_SM02_mo=timetable(TIME_month,SM02_month);

WDIR_M=TT.WDIR;
WDIR_month=WDIR_M(mo,:);
TT_WDIR_mo=timetable(TIME_month,WDIR_month);

WSPD_M=TT.WSPD;
WSPD_month=WSPD_M(mo,:);
TT_WSPD_mo=timetable(TIME_month,WSPD_month);

WSMX_M=TT.WSMX;
WSMX_month=WSMX_M(mo,:);
TT_WSMX_mo=timetable(TIME_month,WSMX_month);

SRAD_M=TT.SRAD;
SRAD_month=SRAD_M(mo,:);
TT_SRAD_mo=timetable(TIME_month,SRAD_month);

TT_month=[TT_TAIR_mo TT_DWPT_mo TT_TAIRx_mo TT_TAIRn_mo TT_PRCP_mo TT_RELH_mo TT_PRES_mo ...
          TT_SM02_mo TT_WDIR_mo TT_WSPD_mo TT_WSMX_mo TT_SRAD_mo];

end

File: CCcodes_python/driverFULL_monthly.py
"""
driverFULL_monthly.py

A Python implementation of the MATLAB script driverFULL_monthly.m.
This script handles monthly data processing and preparation for climate indices calculations.

Functions:
- Loads monthly meteorological data
- Processes data into usable structure
- Creates year and month groupings
- Prepares data for subsequent climate indices calculations

Author: Converted from MATLAB
Date: 2025-03-24
"""

import numpy as np
import pandas as pd
import os
from pathlib import Path
from datetime import datetime
import h5py

def get_data_month(tt, month_mask, time):
    """
    Extract monthly data for all variables
    
    Parameters:
    -----------
    tt : DataFrame
        Full dataset
    month_mask : array-like
        Boolean mask for the month
    time : Series
        Timestamps
    
    Returns:
    --------
    tt_month : DataFrame
        Monthly data
    """
    time_month = time[month_mask]
    
    # List of variables to process
    variables = [
        'TAIR', 'DWPT', 'TAIRx', 'TAIRn', 'PRCP', 'RELH', 
        'PRES', 'SM02', 'WDIR', 'WSPD', 'WSMX', 'SRAD'
    ]
    
    # Create a new DataFrame for monthly data
    data_dict = {'TIME_month': time_month}
    
    # Extract each variable for the month
    for var in variables:
        if var in tt.columns:
            data_dict[var + '_month'] = tt[var][month_mask].values
    
    # Create DataFrame and set index
    tt_month = pd.DataFrame(data_dict)
    tt_month.set_index('TIME_month', inplace=True)
    
    return tt_month

def process_monthly_data(site='HCKM', file_name=None):
    """
    Process monthly meteorological data for climate indices calculation
    
    Parameters:
    -----------
    site : str
        Site code
    file_name : str, optional
        Input data file name. If None, uses a default pattern.
    
    Returns:
    --------
    my_struct_m : dict
        Dictionary containing processed monthly data
    """
    print(f"Processing monthly data for site {site}")
    
    # Default file name if not provided
    if file_name is None:
        file_name = f'01-Nov-2009_31-Jul-2023_{site}_daily.csv'
    
    # Base directory
    base_dir = Path('/Users/erappin/Documents/Mesonet/ClimateIndices/cliSITES/')
    site_dir = base_dir / site
    
    # Load threshold file
    thresh_file = site_dir / f"{site}_CLIthresh_daily.csv"
    try:
        thresh_data = pd.read_csv(thresh_file)
        print(f"Loaded threshold data from {thresh_file}")
    except FileNotFoundError:
        print(f"Threshold file not found: {thresh_file}")
        thresh_data = None
    
    # Load meteorological data
    data_file = base_dir / file_name
    try:
        # Try to load as CSV first
        df = pd.read_csv(data_file, parse_dates=['TimestampCollected'])
        print(f"Loaded data from {data_file}")
    except (FileNotFoundError, pd.errors.ParserError):
        try:
            # If CSV fails, try as MATLAB file
            import scipy.io as sio
            mat_file = str(data_file).replace('.csv', '.mat')
            mat_data = sio.loadmat(mat_file)
            # Extract data from the MATLAB structure
            tt_daily = mat_data['TT_dailyMES']
            # Convert to DataFrame (structure depends on MATLAB file)
            if isinstance(tt_daily, np.ndarray) and hasattr(tt_daily, 'dtype') and tt_daily.dtype.names:
                df = pd.DataFrame()
                for name in tt_daily.dtype.names:
                    df[name] = tt_daily[name].flatten()
                # Convert timestamp to datetime
                if 'TimestampCollected' in df.columns:
                    df['TimestampCollected'] = pd.to_datetime(df['TimestampCollected'])
            else:
                print("Unexpected format in MAT file")
                return None
            print(f"Loaded data from {mat_file}")
        except Exception as e:
            print(f"Error loading data: {e}")
            return None
    
    # Extract time information
    time = df['TimestampCollected']
    s_time = time.iloc[0]
    e_time = time.iloc[-1]
    
    # Get first and last years
    s_year, s_month, s_day = s_time.year, s_time.month, s_time.day
    e_year, e_month, e_day = e_time.year, e_time.month, e_time.day
    
    # Create date range for complete years
    date_values = pd.date_range(
        start=datetime(s_year, 1, 1),
        end=datetime(e_year, 12, 31),
        freq='D'
    )
    
    # Ensure data has complete date range (fill with zeros for missing dates)
    df_complete = df.set_index('TimestampCollected').reindex(date_values, fill_value=0)
    df_complete.reset_index(inplace=True)
    df_complete.rename(columns={'index': 'TimestampCollected'}, inplace=True)
    
    # Extract actual time range used
    time_full = df_complete['TimestampCollected']
    is_d = time_full[time_full == s_time].index[0] if np.any(time_full == s_time) else 0
    ie_d = time_full[time_full == e_time].index[0] if np.any(time_full == e_time) else len(time_full) - 1
    
    # Get unique years
    all_years = np.unique(time_full.dt.year)
    
    # Set up structure for monthly data
    months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
    variables = [
        "TAIR_month", "DWPT_month", "TAIRx_month", "TAIRn_month", "PRCP_month", "RELH_month",
        "PRES_month", "SM02_month", "WDIR_month", "WSPD_month", "WSMX_month", "SRAD_month"
    ]
    
    my_struct_m = {
        'year': all_years,
        'month': months,
        'var': variables,
        'data': [[] for _ in range(len(months))],
        'years': [[] for _ in range(len(months))]
    }
    
    # Process data by month
    for i, month in enumerate(range(1, 13)):
        # Create month mask
        month_mask = time_full.dt.month == month
        month_data = df_complete[month_mask]
        
        # Group by year for this month
        year_groups = time_full[month_mask].dt.year
        unique_years = np.unique(year_groups)
        
        # Process each year for this month
        for j, year in enumerate(unique_years):
            # Create mask for this month and year
            year_mask = year_groups == year
            
            # Get data for this month and year
            tt_month = get_data_month(df_complete, month_mask & (time_full.dt.year == year), time_full)
            
            # Store in structure
            if j >= len(my_struct_m['data'][i]):
                my_struct_m['data'][i].append(tt_month)
                my_struct_m['years'][i].append(year)
            else:
                my_struct_m['data'][i][j] = tt_month
                my_struct_m['years'][i][j] = year
    
    # Save data to file
    output_dir = site_dir / 'processed'
    output_dir.mkdir(exist_ok=True, parents=True)
    output_h5 = output_dir / f"{site}_monthly_processed.h5"
    
    with h5py.File(output_h5, 'w') as f:
        # Store metadata
        f.create_dataset('years', data=all_years)
        f.attrs['months'] = np.array(months, dtype='S')
        f.attrs['variables'] = np.array(variables, dtype='S')
        
        # Create group for data
        data_group = f.create_group('data')
        
        # Store data for each month
        for i, month in enumerate(months):
            month_group = data_group.create_group(month)
            
            for j, year in enumerate(my_struct_m['years'][i]):
                if j < len(my_struct_m['data'][i]):
                    year_group = month_group.create_group(str(year))
                    df = my_struct_m['data'][i][j]
                    
                    if df is not None and not df.empty:
                        # Store timestamp
                        timestamps = df.index.strftime('%Y-%m-%dT%H:%M:%S').values
                        year_group.create_dataset('timestamps', data=np.array(timestamps, dtype='S'))
                        
                        # Store variables
                        for var in df.columns:
                            year_group.create_dataset(var, data=df[var].values)
    
    print(f"Saved processed monthly data to {output_h5}")
    
    return my_struct_m

if __name__ == "__main__":
    # Example usage
    site = 'HCKM'
    data = process_monthly_data(site)
